src/text-editor-presenter.coffee

```diff
@@ -298,50 +302,46 @@ class TextEditorPresenter
     @state.content.backgroundColor = if @model.isMini() then null else @backgroundColor
     @state.content.placeholderText = if @model.isEmpty() then @model.getPlaceholderText() else null
 
-  updateLinesState: ->
+  updateTilesState: ->
     return unless @startRow? and @endRow? and @lineHeight?
 
-    visibleLineIds = {}
-    row = @startRow
-    while row < @endRow
-      line = @model.tokenizedLineForScreenRow(row)
-      unless line?
-        throw new Error("No line exists for row #{row}. Last screen row: #{@model.getLastScreenRow()}")
-
-      visibleLineIds[line.id] = true
-      if @state.content.lines.hasOwnProperty(line.id)
-        @updateLineState(row, line)
+    linesPerTile = Math.floor(@height / @lineHeight / @tileCount)
+    linesPerTile = Math.max(1, linesPerTile)
+
+    startIndex = Math.max(
+      0, Math.floor(@startRow / linesPerTile) - @tileOverdrawMargin
+    )
+    endIndex = Math.min(
+      Math.ceil(@model.getScreenLineCount() / linesPerTile),
+      Math.ceil(@endRow / linesPerTile) + @tileOverdrawMargin
+    )
+
+    visibleTiles = {}
+    for index in [startIndex...endIndex]
+      presenter = @linesPresentersByTileIndex[index] ?= new LinesPresenter(@)
```

Comment:
Is it too nasty to just build out the tile in a subroutine of the current presenter?